### PR TITLE
udate difficulty calculation

### PR DESCRIPTION
**Pull request by smtmfft**

### CODE REVIEW

File: lib/src/taiko/protocol_instance.rs [hunk 1]
```diff
@@ -1,17 +1,34 @@
 use anyhow::Result;
+use ethers_core::{
+    abi::{encode_packed, Token},
+    types::U256 as ethU256,
```
Comment (johntaiko) [inline lib/src/taiko/protocol_instance.rs]:
```suggestion
    types::U256 as EthU256,
```

File: lib/src/taiko/protocol_instance.rs [hunk 2]
```diff
@@ -25,19 +42,16 @@ pub fn assemble_protocol_instance(extra: &TaikoExtra, header: &Header) -> Result
...
+        extra.l1_next_block.mix_hash.unwrap_or_default().0.into()
     } else {
-        from_ethers_u256(extra.l1_next_block.difficulty)
+        extra.l1_next_block.difficulty
     };
-    let difficulty = block_hash_h256
-        ^ (prevrando
-            * U256::from(header.number)
-            * U256::from(extra.l1_next_block.number.unwrap_or_default().as_u64()));
+    let difficulty = calc_difficulty(
```
Comment (johntaiko) [inline lib/src/taiko/protocol_instance.rs]:
```suggestion
    use use alloy_sol_types::SolValue;
    let difficulty = keccak::keccak(
                (
                    prevrando,
                    header.number,
                    extra.l1_next_block.number,
                )
                    .abi_encode(),
     );
```

Comment (smtmfft) [reply lib/src/taiko/protocol_instance.rs]: abi_encode is not packed encode??

Comment (johntaiko) [reply lib/src/taiko/protocol_instance.rs]: It can be easier to use the `SolValue` trait to get an ABI encoded value.

Comment (smtmfft) [reply lib/src/taiko/protocol_instance.rs]: ok, let try to switch to SolValue

File: lib/src/taiko/protocol_instance.rs [hunk 3]
```diff
@@ -25,19 +42,16 @@ pub fn assemble_protocol_instance(extra: &TaikoExtra, header: &Header) -> Result
...
     let deposits_hash = deposits_hash(&deposits);
     let extra_data = string_to_bytes32(&header.extra_data);
-    //   meta.difficulty = meta.blobHash ^ bytes32(block.prevrandao * b.numBlocks *
-    // block.number);
-    let block_hash = tx_list_hash;
-    let block_hash_h256: U256 = block_hash.into();
-    let prevrando = if cfg!(feature = "pos") {
-        from_ethers_h256(extra.l1_next_block.mix_hash.unwrap_or_default()).into()
+    let prevrando: ethU256 = if cfg!(feature = "pos") {
+        extra.l1_next_block.mix_hash.unwrap_or_default().0.into()
```
Comment (johntaiko) [inline lib/src/taiko/protocol_instance.rs]: Need to revert this if use the `SolValue` to get an `abi_encode` value,

Comment (smtmfft) [reply lib/src/taiko/protocol_instance.rs]: Yes, it's compatible with previous data type, will change later.

File: lib/src/taiko/protocol_instance.rs (hunk 2 above)
Comment (johntaiko) [reply lib/src/taiko/protocol_instance.rs]: > abi_encode is not packed encode??

Just for example. It also has the `abi_encode_packed` method

File: lib/src/taiko/protocol_instance.rs [hunk 4]
```diff
@@ -1,17 +1,36 @@
...
+    TxHash,
 };
+use alloy_sol_types::SolValue;
 
 use crate::taiko::host::TaikoExtra;
 
+
+fn calc_difficulty(prevrando: EthU256, num_blocks: u64, block_num: EthU256) -> [u8; 32] {
+    // meta.difficulty = keccak256(abi.encodePacked(block.prevrandao, b.numBlocks, block.number));
+    let prevrando_bytes: [u8; 32] = prevrando.into();
```
Comment (johntaiko) [inline lib/src/taiko/protocol_instance.rs]: No need to transfers between types because `SolValue` has implemented for primitive types and alloy types

Comment (smtmfft) [reply lib/src/taiko/protocol_instance.rs]: Not this one, it's a quick test before I switch back to zeth data type. Just submitted.

**Pull request merged**